widgets/linechart/value_formatter.go

```diff
@@ -0,0 +1,146 @@
+// Copyright 2019 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package linechart
+
+// value_formatter.go provides common implementations of ValueFormatter that can be
+// used with the YAxisFormattedValues() LineChart option.
+
+import (
+	"fmt"
+	"math"
+	"strings"
+	"time"
+)
+
+// durationSingleUnitPrettyFormat returns the pretty format in one single
+// unit for a time.Duration, the different returned unit formats
+// are: nanoseconds, microseconds, milliseconds, seconds, minutes
+// hours, days.
+func durationSingleUnitPrettyFormat(d time.Duration, decimals int) string {
+	// Check if the duration is less than 0.
+	prefix := ""
+	if d < 0 {
+		prefix = "-"
+		d = time.Duration(math.Abs(d.Seconds()) * float64(time.Second))
+	}
+
+	switch {
+	// Nanoseconds.
+	case d.Nanoseconds() < 1000:
+		dFmt := prefix + "%dns"
+		return fmt.Sprintf(dFmt, d.Nanoseconds())
+	// Microseconds.
+	case d.Seconds()*1000*1000 < 1000:
+		dFmt := prefix + suffixDecimalFormat(decimals, "µs")
+		return fmt.Sprintf(dFmt, d.Seconds()*1000*1000)
+	// Milliseconds.
+	case d.Seconds()*1000 < 1000:
+		dFmt := prefix + suffixDecimalFormat(decimals, "ms")
+		return fmt.Sprintf(dFmt, d.Seconds()*1000)
+	// Seconds.
+	case d.Seconds() < 60:
+		dFmt := prefix + suffixDecimalFormat(decimals, "s")
+		return fmt.Sprintf(dFmt, d.Seconds())
+	// Minutes.
+	case d.Minutes() < 60:
+		dFmt := prefix + suffixDecimalFormat(decimals, "m")
+		return fmt.Sprintf(dFmt, d.Minutes())
+	// Hours.
+	case d.Hours() < 24:
+		dFmt := prefix + suffixDecimalFormat(decimals, "h")
+		return fmt.Sprintf(dFmt, d.Hours())
+	// Days.
+	default:
+		dFmt := prefix + suffixDecimalFormat(decimals, "d")
+		return fmt.Sprintf(dFmt, d.Hours()/24)
+	}
+}
+
+func suffixDecimalFormat(decimals int, suffix string) string {
+	suffix = strings.ReplaceAll(suffix, "%", "%%") // safe `%` character for fmt.
```

Comment:
Looks like this was added in Go 1.12. We are trying to support the last three versions which currently includes Go 1.10 and 1.11.

https://travis-ci.org/mum4k/termdash/builds/529547980

Can we find a solution compatible with the older Go versions? One option (what I have done in the past) is to simply copy the implementation of the missing strings.ReplaceAll into the Termdash repo with a note that it can be removed when we drop support for Go 1.10 and 1.11.

We could create our own strings package under the internal directory:

https://github.com/mum4k/termdash/tree/master/internal

Reply:
After all the changes I broke the compatibility In the last change haha!

I think that we can fallback to the old known `Replace(s, old, new, -1)`, that is what it does the `ReplaceAll` internally. I don't think that is worth creating a new package for this simple statement. I'll push this change for now and if you prefer to have the internal package I'll do the change.

Reply:
Of course, that is much better. Thank you.